compiler/rustc_codegen_llvm/src/gotoc/logging.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+//! Logging for RMC to write structured logs to a log file.
+use crate::btree_map;
+use rustc_serialize::json::*;
+use std::env;
+use std::fs::OpenOptions;
+use std::io::prelude::*;
+use tracing::debug;
+
+pub use crate::{rmc_debug, rmc_log, rmc_warn};
+
+/// A line in the log file.
+struct LogLine {
+    log_type: LogType,
+    message: String,
+}
+
+impl ToJson for LogLine {
+    fn to_json(&self) -> Json {
+        let output = btree_map![
+            ("log_type".to_string(), self.log_type.to_json()),
+            ("message".to_string(), self.message.to_json())
+        ];
+
+        Json::Object(output)
+    }
+}
+
+/// The kind of log message.
+#[derive(Debug)]
+enum LogType {
+    Debug,
+    Log,
+    Warning(WarningType),
+}
+
+impl ToJson for LogType {
+    fn to_json(&self) -> Json {
+        match self {
+            LogType::Debug => {
+                let output = btree_map![("type".to_string(), "DEBUG".to_json())];
+                Json::Object(output)
+            }
+            LogType::Log => {
+                let output = btree_map![("type".to_string(), "LOG".to_json())];
+                Json::Object(output)
+            }
+            LogType::Warning(warning_type) => {
+                let output = btree_map![
+                    ("type".to_string(), "WARNING".to_json()),
+                    ("warning_kind".to_string(), warning_type.to_json())
+                ];
+                Json::Object(output)
+            }
+        }
+    }
+}
+
+/// The kind of warning message.
+#[derive(Debug)]
+pub enum WarningType {
+    Unsupported,
+    Other,
+}
+
+impl ToJson for WarningType {
+    fn to_json(&self) -> Json {
+        format!("{:?}", self).to_json()
+    }
+}
+
+/// Writes a given log line to the log file specified through 
+/// the the environment variable RMC_LOG_FILE. 
+/// If this is not set, continues without failing.
+fn write_to_log_file(log_line: LogLine) {
+    // Still print debug message to user.
+    debug!("RMC: [{:?}]: {}", log_line.log_type, log_line.message);
```

Comment:
What about warning messages? Don't we want to print those to the user even if debug is off

Reply:
Fixed.